jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/MessageInputStream.java

```diff
@@ -193,7 +198,15 @@ public void close()
             l.signal();
         }
 
-        entries.forEach(e -> e.callback.succeeded());
+        boolean demandRequired = false;
+        for (Entry e : entries)
+        {
+            e.callback.succeeded();
+            if (e.frame != null && !e.frame.isFin())
+                demandRequired = true;
+        }
+        if (demandRequired)
+            session.demand();
```

Comment:
I think this is wrong way around.  What if there was no demand and no frames?  shouldn't this be:
```suggestion
        boolean demandRequired = true;
        for (Entry e : entries)
        {
            e.callback.succeeded();
            if (e.frame != null && e.frame.isFin())
                demandRequired = false;
        }
        if (demandRequired)
            session.demand();
```

Reply:
If there were no frames and no demand it means we have reached the end of the message and it is the responsibility of the FrameHandler or application to do the demand.

And if there is no frame here, there may still be demand, in which case demanding here would result in `ReadPendingException`.

In this `MessageSink` we always demand until we reach the end of the message. So if we are to demand here it has to be because we have a non-fin frame.